File: lua/food-chain.lua
local t = {
    "I know an old lady who swallowed a fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
    "I know an old lady who swallowed a spider.\nIt wriggled and jiggled and tickled inside her.\n"
    .. "She swallowed the spider to catch the fly.\n"
    .. "I don't know why she swallowed the fly. Perhaps she'll die.\n",
    "I know an old lady who swallowed a bird.\n"
    .. "How absurd to swallow a bird!\n"
    .. "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
    .. "She swallowed the spider to catch the fly.\n"
    .. "I don't know why she swallowed the fly. Perhaps she'll die.\n",
    "I know an old lady who swallowed a cat.\n"
    .. "Imagine that, to swallow a cat!\n"
    .. "She swallowed the cat to catch the bird.\n"
    .. "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
    .. "She swallowed the spider to catch the fly.\n"
    .. "I don't know why she swallowed the fly. "
    .. "Perhaps she'll die.\n",
    "I know an old lady who swallowed a dog.\n"
    .. "What a hog, to swallow a dog!\n"
    .. "She swallowed the dog to catch the cat.\n"
    .. "She swallowed the cat to catch the bird.\n"
    .. "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
    .. "She swallowed the spider to catch the fly.\n"
    .. "I don't know why she swallowed the fly. "
    .. "Perhaps she'll die.\n",
    "I know an old lady who swallowed a goat.\n"
    .. "Just opened her throat and swallowed a goat!\n"
    .. "She swallowed the goat to catch the dog.\n"
    .. "She swallowed the dog to catch the cat.\n"
    .. "She swallowed the cat to catch the bird.\n"
    .. "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
    .. "She swallowed the spider to catch the fly.\n"
    .. "I don't know why she swallowed the fly. "
    .. "Perhaps she'll die.\n",
    "I know an old lady who swallowed a cow.\n"
    .. "I don't know how she swallowed a cow!\n"
    .. "She swallowed the cow to catch the goat.\n"
    .. "She swallowed the goat to catch the dog.\n"
    .. "She swallowed the dog to catch the cat.\n"
    .. "She swallowed the cat to catch the bird.\n"
    .. "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
    .. "She swallowed the spider to catch the fly.\n"
    .. "I don't know why she swallowed the fly. "
    .. "Perhaps she'll die.\n",
    "I know an old lady who swallowed a horse.\n" .. "She's dead, of course!\n",
}

local function verse(which)
    return t[which]
end

local function verses(from, to)
    local s = ""
    for i = from, to do
        s = s .. verse(i) .. "\n"
    end
    return s
end

local function sing()
    return verses(1, 8)
end

return {
    verse = verse,
    verses = verses,
    sing = sing,
}

File: elixir/food_chain.ex
defmodule FoodChain do
  @foods ~w(fly spider bird cat dog goat cow horse)

  @doc """
  Generate consecutive verses of the song 'I Know an Old Lady Who Swallowed a Fly'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) do
    start..stop
    |> Enum.map(&recite(&1 - 1))
    |> Enum.join("\n")
  end

  defp recite(v) do
    animal = Enum.at(@foods, v)
    s = "I know an old lady who swallowed a #{animal}.\n"

    s =
      s <>
        case animal do
          "spider" -> "It wriggled and jiggled and tickled inside her.\n"
          "bird" -> "How absurd to swallow a bird!\n"
          "cat" -> "Imagine that, to swallow a cat!\n"
          "dog" -> "What a hog, to swallow a dog!\n"
          "goat" -> "Just opened her throat and swallowed a goat!\n"
          "cow" -> "I don't know how she swallowed a cow!\n"
          "horse" -> "She's dead, of course!\n"
          _ -> ""
        end

    if animal == "horse" do
      s
    else
      v..1//-1
      |> Enum.reduce(s, fn v, acc ->
        acc <>
          case Enum.at(@foods, v - 1) do
            "spider" ->
              "She swallowed the #{Enum.at(@foods, v)} to catch the spider that wriggled and jiggled and tickled inside her.\n"

            u ->
              "She swallowed the #{Enum.at(@foods, v)} to catch the #{u}.\n"
          end
      end)
      |> Kernel.<>(
        "I don't know why she swallowed the #{Enum.at(@foods, 0)}. Perhaps she'll die.\n"
      )
    end
  end
end
